cmd/history.go

```diff
@@ -0,0 +1,115 @@
+	"github.com/pulumi/pulumi/pkg/diag/colors"
+	"github.com/pulumi/pulumi/pkg/util/cmdutil"
+)
+
+func newHistoryCmd() *cobra.Command {
+	var stack string
+	var outputJSON bool // Requires PULUMI_DEBUG_COMMANDS
+	var cmd = &cobra.Command{
+		Use:        "history",
+		Aliases:    []string{"hist"},
+		SuggestFor: []string{"updates"},
+		Short:      "Update history for a stack",
+		Long: "Update history for a stack\n" +
+			"\n" +
+			"This command lists data about previous updates for a stack.",
+		Args: cmdutil.NoArgs,
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+
+			opts := display.Options{
+				Color: cmdutil.GetGlobalColorization(),
+			}
+
+			s, err := requireStack(stack, false, opts, false /*setCurrent*/)
+
+			if err != nil {
+				return err
+			}
+
+			b := s.Backend()
+
+			updates, err := b.GetHistory(commandContext(), s.Ref())
+			if err != nil {
+				return errors.Wrap(err, "getting history")
+			}
+
+			if outputJSON {
+				b, err := json.MarshalIndent(updates, "", "    ")
+				if err != nil {
+					return err
+				}
+				fmt.Println(string(b))
+			} else {
+				printUpdateHistory(updates, opts)
+			}
+			return nil
+		}),
+	}
+	cmd.PersistentFlags().StringVarP(
+		&stack, "stack", "s", "",
+		"Choose an stack other than the currently selected one")
+	// pulumi/issues/496 tracks adding a --format option across all commands. Rather than expose a partial solution
+	// for just `history`, we put the JSON flag behind an env var so we can use in tests w/o making public.
+	if cmdutil.IsTruthy(os.Getenv("PULUMI_DEBUG_COMMANDS")) {
+		cmd.PersistentFlags().BoolVar(&outputJSON, "output-json", false, "Output stack history as JSON")
+	}
+	return cmd
+}
+func printUpdateHistory(updates []backend.UpdateInfo, opts display.Options) {
+
+	if len(updates) == 0 {
+		fmt.Println("Stack has never been updated")
+		return
+	}
+
+	for _, update := range updates {
+
+		fmt.Print(
+			opts.Color.Colorize(
+				fmt.Sprintf("%s%s%s\n", colors.Green, "-----------------------", colors.Reset)))
+
+		fmt.Printf("  UpdateKind: %v \n  Status: %v  m: %v \n", update.Kind, update.Result, update.Message)
+		fmt.Print(
+			opts.Color.Colorize(
+				fmt.Sprintf("  %s%s+%v%s%s%s-%v%s%s%s~%v%s%s%s %v%s",
+					colors.GreenBg, colors.Black, update.ResourceChanges["create"], colors.Reset,
+					colors.RedBg, colors.Black, update.ResourceChanges["delete"], colors.Reset,
+					colors.YellowBg, colors.Black, update.ResourceChanges["update"], colors.Reset,
+					colors.BlueBg, colors.Black, update.ResourceChanges["same"], colors.Reset)))
+
+		tStart := time.Unix(update.StartTime, 0)
+		tCreated := humanize.Time(tStart)
+		tEnd := time.Unix(update.EndTime, 0)
+		duration := tEnd.Sub(tStart)
+
+		fmt.Printf("  Updated %s took %s\n", tCreated, duration)
+
+		if len(update.Environment) != 0 {
+			fmt.Printf("    Github-Login: %s\n", update.Environment["github.login"])
+			fmt.Printf("    Github-Committer: %s <%s>\n",
```

Comment:
The login is getting printed from "github.login"
As for the second line I tried to mimic git log where you have:
username <person@email.com> all on one line

Do you recommend something else? Or only printing a subset of these?

Reply:
I wonder whether we should just print all key/value pairs in the environment. Especially with https://github.com/pulumi/pulumi/issues/2144, we intend to make this extensible, so it seems nice to just print out the full set of tags captured. I think this will just auto-magically include CI variables, for instance, which are nice to print out and include. This would also make the CLI easier to maintain because we don't need to remember to come back and add them each time we add a new tag.

(Note that if we do this, we'll want to sort the keys so the map is printed in deterministic order, since Go intentionally randomizes map enumeration order.)

Reply:
![environmentmap](https://user-images.githubusercontent.com/8211780/49785216-abc6d100-fcd4-11e8-9ea6-8fbcef041221.png)